File: src/backend/SmartCrops.Core/Dashboard/DashboardPresets.cs
using System.Diagnostics.CodeAnalysis;

namespace SmartCrops.Core.Dashboard;

/// <summary>
/// One block of a preset: its footprint and whether the level hides it by
/// default. Order comes from the enclosing list, not from a field.
/// </summary>
/// <param name="Key">One of <see cref="DashboardLayout.Blocks"/>.</param>
/// <param name="Size">One of <see cref="DashboardLayout.Sizes"/>.</param>
/// <param name="Hidden">True when the level does not show the block by default.</param>
public record DashboardPresetBlock(string Key, string Size, bool Hidden);

/// <summary>
/// SMA-336 — the three level presets, verbatim from the frozen design
/// (<c>_spec.md</c> § 8): « Novice : Météo Moyen, Jardins Moyen, Conseils Petit,
/// Ce mois-ci Petit », « Jardinier : Météo Moyen, Jardins Grand, Conseils Moyen,
/// Ce mois-ci Moyen, À faire Moyen, Compteurs Moyen », « Expert : les huit
/// widgets en Grand ». Since SMA-448 (lot F1) they are also the WIDGETS each
/// formula has, and <see cref="FormulaCatalog"/> serves them.
///
/// <para>Every preset lists every block its level permits: a block a level does
/// not show is present and <c>Hidden</c>, so the Customize gallery can offer it
/// back without inventing an entry. And ONLY those (SMA-437 lot 1, PR B, step
/// B1 — pre-flight D4): the blocks of a level ARE those of its preset
/// (<see cref="Permits"/>), and the Key figures band is the Expert's alone —
/// « en tête du preset Expert », in its one size, the Full width (D8). This is
/// also the fallback the API serves when a user has no saved layout, or one
/// stored under an unknown schema version.</para>
///
/// <para>Byte-identical to the client's <c>constants/dashboardPresets.ts</c>,
/// and checked so: both sides compare their presets to
/// <c>src/frontend/src/constants/dashboardLayout.reference.json</c> (PR #287,
/// fix round 1, S2).</para>
/// </summary>
public static class DashboardPresets
{
    private const string S = DashboardLayout.Sizes.Small;
    private const string M = DashboardLayout.Sizes.Medium;
    private const string L = DashboardLayout.Sizes.Large;
    private const string W = DashboardLayout.Sizes.Wide;

    private static readonly IReadOnlyList<DashboardPresetBlock> NovicePreset =
    [
        new(DashboardLayout.Blocks.Weather, M, false),
        new(DashboardLayout.Blocks.Gardens, M, false),
        new(DashboardLayout.Blocks.Tips, S, false),
        new(DashboardLayout.Blocks.Month, S, false),
        new(DashboardLayout.Blocks.Todo, M, true),
        new(DashboardLayout.Blocks.Counters, M, true),
        new(DashboardLayout.Blocks.Stats, L, true),
        new(DashboardLayout.Blocks.Harvest, L, true),
    ];

    /// <summary>
    /// Without Statistics since the formulas (SMA-448, lot F1 — R1, V3-01:
    /// « Les statistiques — Non · Non · Oui », « retiré au Jardinier »): a
    /// preset lists the blocks its formula HAS, so the Gardener cannot bring
    /// Statistics back from its gallery, nor save it. Récolte stays, hidden, as
    /// before (Alexandre, 26/09, question 3).
    /// </summary>
    private static readonly IReadOnlyList<DashboardPresetBlock> GardenerPreset =
    [
        new(DashboardLayout.Blocks.Weather, M, false),
        new(DashboardLayout.Blocks.Gardens, L, false),
        new(DashboardLayout.Blocks.Tips, M, false),
        new(DashboardLayout.Blocks.Month, M, false),
        new(DashboardLayout.Blocks.Todo, M, false),
        new(DashboardLayout.Blocks.Counters, M, false),
        new(DashboardLayout.Blocks.Harvest, L, true),
    ];

    /// <summary>
    /// Written by hand since the band (pre-flight D8), where it was derived from
    /// the keys: the band first, in the Full width, then the eight in Large.
    /// </summary>
    private static readonly IReadOnlyList<DashboardPresetBlock> ExpertPreset =
    [
        new(DashboardLayout.Blocks.KeyFigures, W, false),
        new(DashboardLayout.Blocks.Weather, L, false),
        new(DashboardLayout.Blocks.Gardens, L, false),
        new(DashboardLayout.Blocks.Tips, L, false),
        new(DashboardLayout.Blocks.Month, L, false),
        new(DashboardLayout.Blocks.Todo, L, false),
        new(DashboardLayout.Blocks.Counters, L, false),
        new(DashboardLayout.Blocks.Stats, L, false),
        new(DashboardLayout.Blocks.Harvest, L, false),
    ];

    /// <summary>
    /// The preset for a level. An unknown level falls back to
    /// <see cref="DashboardLayout.DefaultLevel"/> rather than throwing: a stored
    /// document must never be able to make a read fail.
    /// </summary>
    public static IReadOnlyList<DashboardPresetBlock> For(string? level) => level switch
    {
        DashboardLayout.Levels.Novice => NovicePreset,
        DashboardLayout.Levels.Expert => ExpertPreset,
        _ => GardenerPreset,
    };

    /// <summary>
    /// Whether <paramref name="level"/> has the block <paramref name="key"/> at
    /// all — whether its preset lists it (pre-flight D4). The minimal right of a
    /// level the controller checks: a block it does not permit is refused on
    /// write and dropped on read. An unknown level reads as the default one, as
    /// <see cref="For"/> does. The client's twin is <c>permitsBlock</c>.
    /// </summary>
    public static bool Permits(string? level, string key) =>
        For(level).Any(block => block.Key == key);

    /// <summary>
    /// True when the level is one this server knows. Annotated so the compiler
    /// narrows <paramref name="level"/> to non-null at the call site.
    /// </summary>
    public static bool IsKnownLevel([NotNullWhen(true)] string? level) =>
        level is not null && DashboardLayout.Levels.All.Contains(level);
}

File: src/frontend/src/constants/dashboardPresets.ts
import type {
  DashboardBlock,
  DashboardBlockKey,
  DashboardLevel,
} from '../types/Dashboard';

/**
 * SMA-336 — the three level presets, transcribed from the frozen design
 * (`_spec.md` § 8) and kept byte-identical to the server's
 * `DashboardPresets.cs`: the client must be able to tell « this layout is the
 * preset » without a round-trip, and « réinitialiser au niveau X » must produce
 * exactly what a fresh account would receive. Checked so since PR #287, fix
 * round 1 (S2): both sides compare their presets to
 * `dashboardLayout.reference.json`.
 *
 * Every preset lists EVERY widget its formula permits — the ones a level does
 * not show are present and `hidden`, so the Customize gallery has something to
 * offer back instead of inventing an entry. And ONLY those (SMA-437 lot 1,
 * PR B, step B1 — pre-flight D4): the blocks of a formula ARE those of its
 * preset, the rule the server refuses a write by and the client and the server
 * drop a stored block by. The Key figures band is the Expert's alone (V3-01),
 * so the Novice and the Gardener list the eight others.
 */
const S = 'small' as const;
const M = 'medium' as const;
const L = 'large' as const;
const W = 'wide' as const;

const NOVICE: DashboardBlock[] = [
  { key: 'weather', size: M, hidden: false },
  { key: 'gardens', size: M, hidden: false },
  { key: 'tips', size: S, hidden: false },
  { key: 'month', size: S, hidden: false },
  { key: 'todo', size: M, hidden: true },
  { key: 'counters', size: M, hidden: true },
  { key: 'stats', size: L, hidden: true },
  { key: 'harvest', size: L, hidden: true },
];

/**
 * Without Statistics since the formulas (SMA-448, lot F1 — R1, V3-01: « Les
 * statistiques — Non · Non · Oui »): the Gardener cannot bring it back from its
 * gallery, and the server refuses it shown. Récolte stays, hidden (Alexandre,
 * 26/09, question 3).
 */
const GARDENER: DashboardBlock[] = [
  { key: 'weather', size: M, hidden: false },
  { key: 'gardens', size: L, hidden: false },
  { key: 'tips', size: M, hidden: false },
  { key: 'month', size: M, hidden: false },
  { key: 'todo', size: M, hidden: false },
  { key: 'counters', size: M, hidden: false },
  { key: 'harvest', size: L, hidden: true },
];

/**
 * Written by hand since the band (pre-flight D8), where it was derived from the
 * keys: « Chiffres clés en tête » (contract § 3.3 [A], § 4.5), in the Full
 * width — its one size — then the eight widgets in Large, as before.
 */
const EXPERT: DashboardBlock[] = [
  { key: 'keyfigures', size: W, hidden: false },
  { key: 'weather', size: L, hidden: false },
  { key: 'gardens', size: L, hidden: false },
  { key: 'tips', size: L, hidden: false },
  { key: 'month', size: L, hidden: false },
  { key: 'todo', size: L, hidden: false },
  { key: 'counters', size: L, hidden: false },
  { key: 'stats', size: L, hidden: false },
  { key: 'harvest', size: L, hidden: false },
];

export const DEFAULT_DASHBOARD_LEVEL: DashboardLevel = 'gardener';

const PRESETS: Record<DashboardLevel, DashboardBlock[]> = {
  novice: NOVICE,
  gardener: GARDENER,
  expert: EXPERT,
};

/**
 * A FRESH copy of a level's preset — the caller mutates its result (drag,
 * resize, hide) and must never reach the module-level arrays.
 */
export function presetFor(level: DashboardLevel): DashboardBlock[] {
  return PRESETS[level].map((block) => ({ ...block }));
}

/**
 * Whether `level` has the widget `key` at all — whether its preset lists it
 * (pre-flight D4). The twin of the server's `DashboardPresets.Permits`.
 */
export function permitsBlock(level: DashboardLevel, key: DashboardBlockKey): boolean {
  return PRESETS[level].some((block) => block.key === key);
}

/**
 * True when the layout no longer matches its level's preset — the « · ajustée »
 * suffix of the level chip. Compared on the three things the user can change in
 * the Edit mode: ORDER (index by index), size and visibility. `options` is
 * deliberately out: PR 1/5 writes none, and a future widget option must not
 * silently relabel a layout the user never rearranged.
 */
export function isAdjusted(
  blocks: DashboardBlock[],
  level: DashboardLevel
): boolean {
  const preset = PRESETS[level];
  if (blocks.length !== preset.length) return true;
  return blocks.some(
    (block, index) =>
      block.key !== preset[index]!.key ||
      block.size !== preset[index]!.size ||
      block.hidden !== preset[index]!.hidden
  );
}
